sys/net/transport_layer/ng_udp/ng_udp.c

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ * @ingroup     sys_net_udp
+ *
+ * @file
+ * @brief       UDP implementation
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @}
+ */
+
+#include <stdint.h>
+
+#include "kernel.h"
+#include "msg.h"
+#include "thread.h"
+#include "net/ng_udp.h"
+#include "net/ng_nettype.h"
+#include "net/ng_netapi.h"
+#include "net/ng_netreg.h"
+#include "net/ng_pktbuf.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/**
+ * @brief   Save the UDP's thread PID for later reference
+ */
+static kernel_pid_t _pid = KERNEL_PID_UNDEF;
+
+/**
+ * @brief   Allocate memory for the UDP thread's stack
+ */
+static char _stack[NG_UDP_STACK_SIZE];
+
+/**
+ * @brief   Calculate the UDP checksum dependent on the network protocol
+ *
+ * @param[in] pkt       pointer to the packet in the packet buffer
+ *
+ * @return              the checksum of the pkt
+ * @return              0 on error
+ */
+uint16_t _calc_checksum(pktsnip_t *pkt)
+{
+    uint16_t cksum = 0;
+    /* TODO: calculate checksum depending on used IP version */
+    return cksum;
+}
+
+void _receive(pktsnip_t *pkt)
+{
+    ng_pktsnip_t *payload;
+    ng_pktsnip_t *udp = pkt;
+    ng_udp_hdr_t *hdr;
+    msg_t msg;
+    ng_netreg_entry_t *reciever;
+
+    /* mark UDP header */
+    while (udp && udp->type != NG_NETTYPE_UDP) {
```

Comment:
This whole chunk of code doesn't use the packet buffer as it was intended. You duplicate a lot of data unnecessarily this way + only the `pktsnip_t` you received will have the counter that interests `ng_pktbuf_start_write()` (otherwise hold must iterate through the whole packet). Should be looking something like:

```C
void _receive(pktsnip_t *pkt)
{
    ng_pktsnip_t *udp, *payload;
    ng_udp_hdr_t *hdr;
    msg_t msg;
    ng_netreg_entry_t *reciever;

    pkt = ng_pktbuf_start_write(pkt);
    LL_SEARCH_SCALAR(pkt, udp, type, NG_NETTYPE_UDP);
    
    if (udp == NULL || udp->size < NG_UDP_HDR_LENGTH) {
        DEBUG("udp: received packet which did not contain UDP data\n");
        ng_pktbuf_release(pkt);
    }
    payload = udp->next;   /* We never discussed this, but judging from your code I guess you assume the previous layer marked the UDP packet, so you don't need to mark it's payload yourself was already set by previous. */
    hdr = (ng_udp_hdr_t *)udp->data;
    /* UDP and handover stuff */
}
```

I'm not quite sure if this presetting of headers works. See https://github.com/RIOT-OS/RIOT/pull/2430/files#r24488558.

An alternative would be, that every header is responsible for it's own headers and the preceeding one just sets the type of the payload to `NG_NETTYPE_UNDEF`

```C
void _receive(pktsnip_t *pkt)
{
    ng_pktsnip_t *udp;
    ng_udp_hdr_t *hdr;
    msg_t msg;
    ng_netreg_entry_t *reciever;

    pkt = ng_pktbuf_start_write(pkt);
    LL_SEARCH_SCALAR(pkt, udp, type, NG_NETTYPE_UNDEF);
    udp->type = NG_NETTYPE_UDP;
    ng_pktbuf_add(udp,   /* Mark payload for next header */
                  (void *)(((uint8_t *)udp->data) + NG_UDP_HDR_LENGTH),
                  udp->size - NG_UDP_HDR_LENGTH,
                  NG_NETTYPE_UNDEF);
    hdr = (ng_udp_hdr_t *)udp->data;
    /* UDP and handover stuff */
}
```